auspice_client_customisation/handleDroppedFiles.js

```diff
@@ -265,42 +203,122 @@ async function loadDatasets(dispatch, datasets, narrative) {
  * dataset to load (if present in the drag & dropped files).
  */
 async function parseNarrative(fileText, datasets, logs) {
-  const datasetsForNarrative = {};
-  const blocks = await parseMarkdownNarrativeFile(fileText, parseMarkdown);
-  addEndOfNarrativeBlock(blocks)
-
-  /* link each "block" to a dataset. This is not straightforward, as narrative slides
-  are linked to a URL not a filepath. Auspice uses the function `getDatasetNamesFromUrl` to
-  perform this linking and expects the appropriate dataset to be present in the cache.
-  We therefore modify `datasets` so that these keys exist */
-  logs.push("Linking narrative datasets to dropped JSON datasets.")
-  const prefixesSeen = new Set();
-  for (const block of blocks) {
-    for (const prefix of getDatasetNamesFromUrl(block.dataset)) {
-      if (prefix && !prefixesSeen.has(prefix)) {
-        const filename = convertPrefixToDatasetFilename(prefix);
-        // Must use lowercase filename to find dataset since the datasets
-        // are stored in lowercase within collectDatasets
-        const filenameLower = filename.toLowerCase();
-        if (datasets[filenameLower]) {
-          logs.push(`Narrative slide URL ${prefix} → ${filename}`)
-          datasetsForNarrative[prefix] = datasets[filenameLower];
-        } else {
-          logs.push(`Narrative slide URL ${prefix} expected ${filename} but this wasn't found.`)
+  const slides = await parseMarkdownNarrativeFile(fileText, parseMarkdown);
+  addEndOfNarrativeBlock(slides)
+
+  /**
+   * Each narrative slide specifies a dataset (referenced elsewhere as the 'urlName'
+   * which is the keys of `datasets`)
+   */
+  const referencedUrlNames = Array.from(new Set(
+    slides.map((slide) => getDatasetNamesFromUrl(slide.dataset)).flat()
+  )).filter((name) => !!name);
+
+  /* Delete any keys from `datasets` that aren't needed for this narrative */
+  for (const name of Array.from(Object.keys(datasets))) {
+    if (!referencedUrlNames.includes(name)) {
+      logs.push(`\tDropping dataset for ${name} as it's not referenced in the narrative`);
+      delete datasets[name]
+    }
+  }
+
+  const missingDatasets = referencedUrlNames.filter((name) => !datasets[name]);
+  for (const name of missingDatasets) {
+    logs.push(`\tNarrative references dataset for ${name} but this wasn't provided. Expect errors!`);
+  }
+  if (missingDatasets.length===0) {
+    logs.push(`\tAll ${referencedUrlNames.length} datasets referenced are present`);
+  }
+
+  return slides;
+}
+
+function makeDroppedFile(file) {
+  const filename = file.name;
+  const filenameLower = filename.toLowerCase();
+  let filenameBase = filenameLower;
```

Comment:
filenameLower isn't used beyond this two lines:

```suggestion
  let filenameBase = filename.toLowerCase();
```

Reply:
🤦